tensorflow_similarity/augmenters/barlow.py

```diff
@@ -0,0 +1,148 @@
+        if self.random_execute(0.5):
+            x = tf.image.random_flip_left_right(x)
+        return x
+
+
+class RandomResizedCrop(Augmentation):
+    def __init__(self):
+        super(Augmentation, self).__init__()
+
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        x_dim = tf.shape(x)
+        x_width = tf.cast(x_dim[0], tf.float32) 
+        x_height = tf.cast(x_dim[1], tf.float32)
+
+        print(x_width)
+        rand_size_width = tf.random.uniform(
+            shape=[],
+            minval=int(0.75 * x_width),
+            maxval=int(1 * x_width),
+            dtype=tf.int32,
+        )
+
+        rand_size_height = tf.random.uniform(
+            shape=[],
+            minval=int(0.75 * x_height),
+            maxval=int(1 * x_height),
+            dtype=tf.int32,
+        )
+
+        crop = tf.image.random_crop(x, (x_width, x_height, 3))
+        crop_resize = tf.image.resize(crop, (x_width, x_height))
+        return crop_resize
+
+
+class RandomSolarize(Augmentation):
+    @tf.function
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        if self.random_execute(0.2):
+            # flips abnormally low pixels to abnormally high pixels
+            x = tf.where(x < 10, x, 255 - x)
+        return x
+
+
+class RandomBlur(Augmentation):
+    @tf.function
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        if self.random_execute(0.2):
+            s = np.random.random()
+            return tfa.image.gaussian_filter2d(image=x, sigma=s)
+        return x
+
+
+class BarlowTransformator(keras.Model):
+    def __init__(self):
+        super(BarlowTransformator, self).__init__()
+        self.random_resized_crop = RandomResizedCrop()
+        self.random_flip = RandomFlip()
+        self.random_color_jitter = RandomColorJitter()
+        self.random_blur = RandomBlur()
+        self.random_to_grayscale = RandomToGrayscale()
+        self.random_solarize = RandomSolarize()
+
+    def call(self, x: tf.Tensor) -> tf.Tensor:
+        x = self.random_resized_crop(x)
+        x = self.random_flip(x)
+        x = self.random_color_jitter(x)
+        x = self.random_blur(x)
+        x = self.random_to_grayscale(x)
+        x = self.random_solarize(x)
+
+        x = tf.clip_by_value(x, 0, 1)
+        return x
+
+class BarlowAugmenter(Augmenter):
+  def __init__(self, num_cpu: Optional[int] = os.cpu_count(), seed: Optional[int] = 128):
+    super(Augmenter, self).__init__()
+    self.num_cpu = num_cpu
+    self.seed = seed
+
+  def augment(
+      self, x: Any, y: Any = tf.constant([0]), num_augmentations_per_example: int = 2, is_warmup: bool = True,
+  ) -> List[Any]:
+
+      with tf.device("/cpu:0"):
+          inputs = tf.stack(x)
+          inputs = tf.cast(inputs, dtype="float32") / 255.0
+          views = []
+          augmenter = BarlowTransformator()
+          
+          for _ in range(num_augmentations_per_example):
+              view = (
+                  tf.data.Dataset.from_tensor_slices(x)
```

Comment:
I think we can get the benefit of the parallel calls here using tf.map_fn, otherwise we will incur the cost of creating the dataset on every batch * num_augmentations_per_example.

Reply:
I see, that makes sense why i felt that it was very slow. Previously it had not been enclosed by @function